elf2cfetbl_version.h

```diff
@@ -0,0 +1,46 @@
+/*
+**  GSC-18128-1, "Core Flight Executive Version 6.7"
+**
+**  Copyright (c) 2006-2019 United States Government as represented by
+**  the Administrator of the National Aeronautics and Space Administration.
+**  All Rights Reserved.
+**
+**  Licensed under the Apache License, Version 2.0 (the "License");
+**  you may not use this file except in compliance with the License.
+**  You may obtain a copy of the License at
+**
+**    http://www.apache.org/licenses/LICENSE-2.0
+**
+**  Unless required by applicable law or agreed to in writing, software
+**  distributed under the License is distributed on an "AS IS" BASIS,
+**  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+**  See the License for the specific language governing permissions and
+**  limitations under the License.
+*/
+
+/*
+ *  File: elf2cfetbl_version.h
+ *
+ *  Purpose:
+ *  Provide version identifiers for the ELF to cFE Table Converter.
+ *  See cfe documentation for version and build number and description
+ */
+
+#ifndef ELF2CFETBL_VERSION_H
+#define ELF2CFETBL_VERSION_H
+
+/*
+ * Development build identifiers
+ */
+#define ELF2CFETBL_BASELINE "Bootes"
+#define ELF2CFETBL_BUILDNUMBER 34
+
+/*
+ * Version Macro Definitions
+ */
+#define ELF2CFETBL_MAJOR_VERSION 3
+#define ELF2CFETBL_MINOR_VERSION 1
+#define ELF2CFETBL_REVISION      0
```

Comment:
Should I reset this or keep the old 3.1.5 and hardcode the tag part of the "DEVELOPMENT" string in `elf2cfetbl.c:1284`?

Reply:
@skliper @jphickey

Reply:
It's more clear to me with what you did here.  Since there isn't a v3.1.5 tag.

Reply:
Yeah, my question was do I hardcode the string with "since v3.1.0 tag" or do I make it reference the macros (is that what they're called?) and then set the revision to zero despite the fact that we've been increasing it for a while.

Reply:
Recommend to avoid hardcoding the string - I actually don't think you need the extra string at all.  Can we just display "Bootes-34" at runtime and indicate in the documentation what that means?  Ideally we can just document it once for all components in the framework.

Reply:
Or just call it "v3.1.0+dev34" which I think captures all the same info in a lot fewer words.

Reply:
I also think that, even though we haven't really changed any APIs, we should bump the bootes version number here to 3.2.0 to avoid confusion with the 3.1.x non-releases.

Reply:
I like that format. 

Last part of my question still stands, should we reset the revisions down to zero for the old development builds, retroactively in some sense, now that we're adding the build number.

Reply:
Not sure about that. I'm using the Version numbers to create the string "v3.1.0+dev". If we bump them to 3.2.0 then I'll have to hardcode the "3.1.0" part of the string.

Reply:
Well, I was thinking the string gets assembled from `ELF2CFETBL_xxx` macros - meaning `ELF2CFETBL_MINOR_VERSION would get preemptively bumped to 2 even though there is no tag until final.  This approach could, in theory, be applied to all modules for which we bumped revision on a dev/non-release build.

I was only concerned about the possible weirdness of going from 3.1.5 -> 3.1.0 which seems like a backstep, and then if at some point in the future we actually _do_ need to do a point release, then we can't call it 3.1.1 - we already have that number used.

Reply:
Actually, the long delay in getting stuff "officially tagged" after a new release is ready is reason (I think) that we should pick an old epoch for build numbers and __not__ attempt to move it with each release.

Reply:
Picking v3.0.0a as the epoch of elf2cfetbl puts the current IC (c0c0f38) at build number 42, and since that is the ultimate answer to everything, it must be right.

Reply:
Yep, I'm concerned as well, I think I'm going to compromise and keep the macros at 3.1.5 but hardcode the string until we get 3.2 released. Once that happens we can start fully using those macros

Reply:
But we already update the macros prior to any "official tag" being created.  Why wouldn't we just update the macro to show be minor version 2 in anticipation of that being the next release?  We don't want any displayed strings to differ (3.1.5 vs. 3.2.0) depending on how they got the info.  There should be just one "right" version to show.

Reply:
For instance in CFE we will be updating the version to 6.8.0 in the version.h file so it reports as such for testing, but the tag is probably still months away, right?  Likewise for any future development we don't want to have to wait until 6.8.0 is officially tagged to get the build numbers right, that could be a while.

Reply:
Also this isn't a one-time issue - we'll have a similar issue every time we prep code for release and it sits awaiting lawyer approval for months before it can be tagged - creating a mismatch between the baseline we are developing against and what the last "official tagged" version is.

Reply:
What if we set all of the macros to zero for development? That way it's clear that people shouldn't pay attention to them and there's no confusion.

Reply:
No, I think the major/minor/revision its still relevant for dev builds.  At the very least, it says what release the development build started from, which is meaningful and helpful.   I think the numbers should just stay the same.

Reply:
We just need a way to identify what is a development build (main/master branch) vs. an official release tag - the latter should not contain the "+dev#" in its reporting.  Such as perhaps a separate/additional snapshot of the build num of the release, and if the current build number and the release build number are equal, then consider it a release and not dev.

Reply:
I think that zeroing out the major, minor, and revision macros sends a clear message that "something is different". Adding a commit that sets these can then be part of the release and tagging process. 

Keeping the version macros populated with either the next version or the previous one and adding comment is another option, but, lets face it, people don't read those. The baseline, codename, and build number and the version string with "dev+" we should have enough to independently identify a development build.

Reply:
The presence of a "+dev#" string in the version output should be the marker that "something is different" - no need to also zero out the version numbers, which the indicator of the most recent numbered version on which your dev branch is based.  That's a manual thing to modify between the mainline and any release branches, and will cause perpetual nuisance merge conflicts every time we need to merge between the branches because it will be seen as a parallel modification on both branches.  No need to go through that - just leave the version numbers as-is, they are helpful to see in both release and engineering/dev versions.

I'm also not sure if we want to even include the BASELINE ("Bootes") in tools like this either - these don't change much aside from minor bugfix revs.  Aquila and whatever is next (C) can likely use the same tool.  If we release a bugfix/patch to an older CFE (e.g. aquila)  we may very well want to include the same/current elf2cfetbl version in that.

Note we can create an aquila or bootes BRANCH, and that's fine because these can point to the same commit.  But I don't think we should bake it into the code for a common utility like this.

Reply:
The idea to set the versions to zero is meant to reduce confusion. We want the version-build information to be obvious to both people who run the code, regardless of how they obtained it, and people who just open `version.h`. If I open `version.h` and see version numbers filled in, I might assume that that's the version number which may NOT be the case. I've tried to put together a picture that kind of explains this. I'll go back to using codename for this example. 

```c
CODENAME "Bootes" /* This indicates compatibility across components */
BASELINE "v3.1.0"      /* Epoch for build number. Probably last tagged MINOR release. */
BUILDNUMBER 34     /* Number of commits since baseline */
```
For cFE this would look like 


|  | 1 | 2 | 3 | 4 | 5 | 6 |
| ------------- | ---------- | ----------- | ------------- | ---------- | ---------- | ---------- |
| codename | Aquila | Bootes | Bootes | C-Constellation | Bootes | C-Constellation | 
| baseline | 6.6? | 6.7.0 | 6.7.0 | 6.7.0  | 6.8.0 | 6.7.0 | 6.8.0 |
| buildnumber | n_A | n_B | m_B  | n_C | p_B | m_C |
| MAJOR,MINOR,REV in version.h | 6.7.0 | 6.7.21 | **6.8.0** |  **6.8.0** | 6.8.x | **6.8.0** |

![cfeversion-01](https://user-images.githubusercontent.com/59618057/86990507-bd02e880-c16a-11ea-85f6-b1c8400b9b23.png)



Notice how for **cases 3, 4, and 6** the version macros in `version.h` are all **6.8.0**. Yes, the baseline is different but people who are not power users, and these is who I see as the target audience for this, might be confused. That's the problem I want to solve. I think we're really close.

Reply:
Thanks for the pic - it illustrates nicely, but still I disagree about having a different "baseline" vs. "version".  The table above of codename/baseline/buildnumber/version combos is quite complex and honestly I don't understand it - why would there be three different versions of "Bootes"?

I'm back to the stance that - for build numbering - we should _NOT_ change the epoch  This is due to the long lag time between being "engineering-complete", thereby starting development on the NEXT version, and having an official tag we can baseline against for build numbering.  Case in point for CFE 6.8.0 - we are just about to finalize this and begin dev work for 7.0.0, but a "v6.8.0" tag likely won't be created for months to come.  We don't want our build numbers to reset halfway through our 7.0.0 dev cycle whenever 6.8.0 is finally tagged.  Nor do we want to reset again if there should be a 6.8.1 patch release, either.  We should be able to refer to "CFE dev build 301" or "elf2cfetbl dev build 42" and have a consistent meaning for that.

The `version.h` file should reflect the _engineering_ version - that is, we update it once the code is feature-complete and ready to _start_ the release cycle.  It may be ahead of the last official tag but should never be behind it.

Likewise the epoch for build numbering doesn't need to be in the C/FSW source code at all, as it is only relevant to the script that computes the build number.  Having this in version.h only makes confusion - as it is NOT related to the current software version at all.  This epoch should only be in a completely different file/script that has nothing to do with the software built-in version reporting mechanism.

Reply:
I'm not a fan of having an "engineering version" and whatever release is and whatever development is.  This just adds to the confusion.  I don't think we should mark the version in version.h as what we plan to release....  This definitely would confuse me as a user... your code says its version X.Y.Z but you haven't released it yet?  Maybe if it said release candidate but that seems like too much complexity also.  Why not just leave it until it's released?  We can make release candidate tags whenever we want that indicate what the release will be.

Reply:
> The presence of a "+dev#" string in the version output should be the marker that "something is different" - no need to also zero out the version numbers, which the indicator of the most recent numbered version on which your dev branch is based. That's a manual thing to modify between the mainline and any release branches, and will cause perpetual nuisance merge conflicts every time we need to merge between the branches because it will be seen as a parallel modification on both branches. No need to go through that - just leave the version numbers as-is, they are helpful to see in both release and engineering/dev versions.
> 
> I'm also not sure if we want to even include the BASELINE ("Bootes") in tools like this either - these don't change much aside from minor bugfix revs. Aquila and whatever is next (C) can likely use the same tool. If we release a bugfix/patch to an older CFE (e.g. aquila) we may very well want to include the same/current elf2cfetbl version in that.
> 
> Note we can create an aquila or bootes BRANCH, and that's fine because these can point to the same commit. But I don't think we should bake it into the code for a common utility like this.

This makes more sense to me... except keep version.h as the last release.  So v6.X.X-dev# for development version, where 6.X.X was the last release that this is based on...  use the names on GitHub releases and keep it to what works together ("Bootes" all work together, "Bootes.1" all work together, etc)?

Reply:
I dislike having an "engineering version" too - but this is the reality we live in when it takes 9-18 months between completing a release from an engineering perspective and being able to publish that as an actual, official release.  To deny that reality in an attempt to keep things "simple" is, I think, more confusing than to accept it and work with it.

As testers are testing it, it should self-report as version "6.8.0" not "6.7.21" or whatever.

We could call it something other than an engineering version, but I think we need to acknowledge that the version we are working on is, in many cases, well ahead of the last official release tag.

Reply:
Testers report related to the release candidate tag "rc-6.8.0", and could include 6.7.0-dev32 as the reported tag.  That makes perfect sense to me.  As soon as you put it as the version number in the code, then you can't change the code without changing the version number.

Reply:
and if "rc-6.8.0" tag has changes, then you just go to rc-6.8.0-1, which would report as 6.7.0-dev33 or whatever.  No breaking any 7150.2C or OPC expectations.

Reply:
Basically there can't be two commits in the wild that report the same complete version string, but are different.

Reply:
They are most definitely NOT testing 6.8.0... because it won't be 6.8.0 until all the headers are updated.  They can test 6.8.0 after it's released, but not before.  They can only test the 6.8.0 release candidate (which is actualy 6.7.0-dev#).  Unless someone wants to try to get agreement from OPC, I strongly recommend this approach.  It is proven to be acceptable by OPC.

Reply:
This honestly makes no sense to me -- as a user, if I checked out the "6.8.0-rc1" tag, I would expect that code to self-report as version 6.8.0 - such as `6.8.0+dev304` - the "dev" suffix is OK as it is an RC not an official release, so that can be expected.  

But if it self-reported as 6.7.something I would think I didn't build it correctly or I was running the wrong code/version.

Isn't that the whole point of the self reporting mechanism?  Why would we want the code to self-report a version that doesn't match the tag?

Reply:
or go from rc-6.8.0 to rc-6.8.0a... whatever, that doesn't really matter as long as it's unique.

Reply:
This is really simple IMO - Once we have approval for an "official" release, we snapshot the build number at that time.  If the current build number matches the release build number, the self-reporting mechanism does _not_ add the "+dev#" suffix, thereby only showing the `6.8.0` and you know you're on the official release.  Simple and easy.

Reply:
I'm sorry it doesn't make sense to you.  It's just a release candidate.  It does match the tag, in that it is 6.7.0-dev33 that we are testing as a release candidate for 6.8.0.

I guarantee we will not be able to make everybody happy.  It's like debating coding style, we can't pick everyone's favorite method.  What we need to do is be OPC and 7150.2C compliant.  We can not pre-mark the code until it's released.

Reply:
> They are most definitely NOT testing 6.8.0.

Right, they are testing `6.8.0+dev304` or whatever.  It should not show `6.8.0` (bare with no dev#) until we have an official release with everything done.

Reply:
With your scheme 6.8.0+dev304 would actually reference to a release candidate (pre 6.8.0 release), yet 6.8.0+dev632 could be post release development on 6.9.0 (post 6.8.0 release)... so very ambiguous as to what the X.Y.Z means if it doesn't always mean the same thing (what the dev version is based on, or whatever).

Reply:
Yeah, I'm solidly against pre-marking as you suggest.  You are changing what the number means with no indication to the user.

Reply:
>  We can not pre-mark the code until it's released.

What?  Haven't we been doing this in `version.h` all along??  3.1.5 (what is currently showing in version.h for this tool) is not a release.  Neither is 6.7.21 for CFE.  Our documentation should be clear that development versions are not releases, as long as we ensure that it has a "dev#" marker in it, it is NOT a release and should not be confused with an official release, and it can self-report accordingly.

Reply:
Nope, that third number was define as the development build.  So the number matched, and it was explicitly documented as not a release.  The meaning of that number never changed (and neither did the major and minor).

Reply:
I had agreement with OPC about major and minor number update as releases, so we were compliant (or I wouldn't have done it).

Reply:
> You are changing what the number means with no indication to the user.

Disagree there...

Yes, `6.8.0+dev304` might be pre-release, then if the official release happens at e.g. `dev502`, therefore `6.8.0+dev632` might be post-release with additional changes beyond the release.  I don't see the issue/confusion with that.  Once the engineering/steering committee team decides we need to cut a new release, we update version.h accordingly (per semantic rules about what we _actually_ changed - major/minor etc) at the same time we INITIATE the process with legal and get the regression test cycle going.  After all, the engineering team tells legal what the next release number _will be_, based on the engineering that went into it and semver rules, not the other way around, right?

Reply:
Can we move on?  I will not agree with the code reporting a number before it's released.  That's a hard no.  I appreciate the inputs, suggestions, but continuing to debate it is not a valuable use of anyone's time.

Reply:
I still don't see how it is permissible/OK to create version tags like 6.8.0-rc1 or 5.0.3-bv before release, which are _clearly_ preemptive major/minor version bumps, but putting the same info in version.h is not OK?  As long as it is self-reports as "rc"/"dev"/whatever to indicate it is not an official release, I don't see the difference.

Clearly you've had some conversations with OPC folks and have an understanding otherwise, but from a regular user perspective, having `version.h` info report its major/minor differently than the tag they pulled is __absolutely__ confusing.

I'll drop the subject as you request.  Having version-confusion among users never stopped us yet, so I guess we'll just continue doing so.